### PR TITLE
gh-98831: rewrite pattern matching opcodes in the instruction definition DSL

### PR DESCRIPTION
**Pull request by iritkatriel**
Separated these out for @brandtbucher to review.

* Issue: gh-98831

### CODE REVIEW

File: Python/bytecodes.c [hunk 1]
```diff
@@ -2093,61 +2093,42 @@ dummy_func(
...
             else if (_PyErr_Occurred(tstate)) {
                 // Error!
-                goto error;
+                ERROR_IF(true, error);
             }
             else {
                 // Failure!
-                SET_TOP(Py_NewRef(Py_None));
+                attrs = Py_NewRef(Py_None);
             }
```
Comment (brandtbucher) [inline Python/bytecodes.c]: The `ERROR_IF(true, error)` inside of a conditional feels sort of strange to me. What do you think of this?

```c
            else {
                ERROR_IF(_PyErr_Occurred(tstate), error);
                attrs = Py_NewRef(Py_None);
            }
```

**Review by brandtbucher — APPROVED** [review]
Thanks! One tiny comment:

**Review by gvanrossum — COMMENTED** [review]
LGTM, but agree with Brandt's suggestion.

File: Python/bytecodes.c (hunk 1 above)
Comment (iritkatriel) [reply Python/bytecodes.c]: TIL that the generator doesn't like trailing comments after an ERROR_IF().

Comment (gvanrossum) [reply Python/bytecodes.c]: Argh, we should fix that. Currently it uses a regex search, which is very lame given that we have an array of tokens available! But moving substitutions to manipulation of the token array is tricky, so improving the regex is probably easier. Example:
```diff
diff --git a/Tools/cases_generator/generate_cases.py b/Tools/cases_generator/generate_cases.py
index d6f9f64ca0..6e433e4cd3 100644
--- a/Tools/cases_generator/generate_cases.py
+++ b/Tools/cases_generator/generate_cases.py
@@ -389,7 +389,7 @@ def write_body(self, out: Formatter, dedent: int, cache_adjust: int = 0) -> None
         assert dedent <= 0
         extra = " " * -dedent
         for line in self.block_text:
-            if m := re.match(r"(\s*)ERROR_IF\((.+), (\w+)\);\s*$", line):
+            if m := re.match(r"(\s*)ERROR_IF\((.+), (\w+)\);\s*(?://.*)?$", line):
                 space, cond, label = m.groups()
                 # ERROR_IF() must pop the inputs from the stack.
                 # The code block is responsible for DECREF()ing them.
```
But there are several other places where this needs to be done (search for `;\s*$`).

Comment (gvanrossum) [reply Python/bytecodes.c]: See [85be319279](https://github.com/gvanrossum/cpython/commit/85be3192793d4d05d949a64671f3cdea18581650).

**Review by gvanrossum — APPROVED** [review]

File: Python/bytecodes.c (hunk 1 above)
Comment (iritkatriel) [reply Python/bytecodes.c]: Ah sorry I missed your comment.

**Review by gvanrossum — APPROVED** [review]
Thanks!

**Pull request merged**